src/main/java/com/hillel/items_exchange/service/impl/CategoryServiceImpl.java

```diff
@@ -0,0 +1,120 @@
+import org.modelmapper.ModelMapper;
+import org.springframework.stereotype.Service;
+
+import static com.hillel.items_exchange.mapper.UtilMapper.*;
+
+@Service
+@RequiredArgsConstructor
+public class CategoryServiceImpl implements CategoryService {
+
+    private final CategoryRepository categoryRepository;
+    private final SubcategoryService subcategoryService;
+    private final ModelMapper modelMapper;
+
+    @Override
+    public List<String> findAllCategoryNames() {
+        return categoryRepository.findAllCategoriesNames();
+    }
+
+    @Override
+    public List<CategoryDto> findAllCategoryDto() {
+        return new ArrayList<>(convertAllTo(categoryRepository.findAll(),
+                CategoryDto.class, ArrayList::new));
+    }
+
+    @Override
+    public Optional<CategoryDto> findCategoryDtoById(long id) {
+        return categoryRepository.findById(id)
+                .map(category -> convertTo(category, CategoryDto.class));
+    }
+
+    @Override
+    public CategoryDto create(CategoryDto categoryDto) {
+        final Category savedCategory = saveCategory(categoryDto);
+        return convertTo(savedCategory, CategoryDto.class);
+    }
+
+    @Override
+    public CategoryDto update(CategoryDto categoryDto) {
+        final Category updatedCategory = saveCategory(categoryDto);
+        return convertTo(updatedCategory, CategoryDto.class);
+    }
+
+    @Override
+    public void removeById(long categoryId) {
+        categoryRepository.deleteById(categoryId);
+    }
+
+    @Override
+    public boolean isCategoryDtoDeletable(long categoryId) {
+        return isCategoryExistsById(categoryId) && isInternalSubcategoriesHaveNotProducts(categoryId);
+    }
+
+    @Override
+    public boolean isCategoryDtoUpdatable(CategoryDto categoryDto) {
+        return isCategoryExistsByIdAndNameOrNotExistsByName(categoryDto.getId(),
+                categoryDto.getName()) && isSubcategoriesExist(categoryDto);
+    }
+
+    @Override
+    public boolean isCategoryDtoValidForCreating(CategoryDto categoryDto) {
+        return isCategoryNameHasNotDuplicate(categoryDto.getName())
+                && categoryDto.getSubcategories().stream().allMatch(this::isSubcategoryIdEqualsZero);
+    }
+
+    private boolean isSubcategoryIdEqualsZero(SubcategoryDto subcategoryDto) {
+        return subcategoryDto.getId() == 0L;
+    }
+
+    private boolean isCategoryExistsById(long categoryId) {
+        return categoryRepository.existsById(categoryId);
+    }
+
+    private boolean isCategoryNameHasNotDuplicate(String name) {
+        return !categoryRepository.existsByNameIgnoreCase(name);
+    }
+
+    private boolean isSubcategoriesExist(CategoryDto categoryDto) {
+        return categoryDto.getSubcategories().stream()
+                .filter(subcategoryDto -> !isSubcategoryIdEqualsZero(subcategoryDto))
+                .allMatch(subcategoryDto -> subcategoryService.isSubcategoryExistsById(subcategoryDto.getId()));
```

Comment:
@rpkyrych suggested us shouldn't check every ID because for each ID will be created SQL request. So you should get list of existed subcategories by their ids or all of subcategories and then manipulate with that list

Reply:
Nice catch, @lubchik-14 !

Reply:
Done